### ISSUE TITLE
[EX-18] task repositories

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#19]
_Peter De Poorter on 2011-01-30T03:58:02Z says:_ 

 <pre>

Definition:

   a task repository is a predefined task list.

Goal:

   to be able to re-use task lists:

    * for repeating activities
    * from other users

Allows very fast task building based on previous experiences
Everyone can create a task repository

Usage examples
*  a task repository for documenting all tasks that are required to install ubuntu. This can be used as base for every user
  that's installing a new ubuntu version
*  a task repository for documenting all tasks that are required to build a new house
*  an ICT company with different technology departments. Each department maintains his task repository.
*  a task repository for storing all the tasks required for releasing taskwarrior :-) 

A project is build by copying specific tasks from the different task repositories

task create repos:<name>
task repos
    shows the available repositories
task switch repos:<x>
    switch to repos <x>.
    x becomes the active task repository on which all the known task commands can be used
task list repos:<x>
    list all the tasks in the repository <x>
task copy repos:<x> proj:<y> 
    copies the proj:<y> tasks from repos:<x> to the active repository
task upload repos:<url>
    uploads the repository to a place (on the internet)
    a central place on the internet could be used for the storage of all task repositories
task download repos:<url>
task connect repos:<path to repos>
task disconnect repos:<path to repos>

</pre>

See also: http://taskwarrior.org/boards/6/topics/show/1017

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2011-01-30T03:58:02Z
Modified: 2017-01-17T02:55:34Z
```

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-02-12T02:06:28Z says:_ 

 I think of this a some kind of task "template".  I'd like to avoid using the term 'repository'.  For me this is synonymous with 'data at a location', and conflicts with the way I'm thinking about the task server.

Do you think we could get this to be smaller than 9 new commands?  Could you elaborate on connect/disconnect?

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-12T11:15:17Z says:_ 

 I really had 'data at a location' in mind. That's the reason why I have choosen the term repository.
It's a bit comparable with git repositories.

Let's me first explain how I work today:

Currently I've created a second task repository manually:

peter(at)dev:~$ ls /home/peter/.task_repository/
completed.data  pending.data  undo.data

I'm using following line in .taskrc_repository:
 data.location=/home/peter/.task_repository

In .bashrc I've defined shortcuts such as:
alias rtsy='task rc:~/.taskrc_repository list proj:system'

The second task repository I call it the reference repository. In the reference repository I've created the tasks that I need to do when I'm installing a system.

When I install a system I manually copy the lists of tasks via vim from the /home/peter/.task_repository/pending.data file to the
/home/peter/.task/pending.data file. I then directly have the lists of tasks that I need to execute in my 'active repository'. I can then start managing them (planning by adding due dates, follow up execution,....) in my active repository.

Above is an example of a locally created repository. If I've done my preparation well others will be interested in using my prepared task list. The idea was then to allow others to clone (such as in git) my repository and sync it when needed (comparable with git pull). 

As you can see my way of working is currently not really a user friendly way of working. With this feature request I hope that together we can find a good approach to obtain the goal in a most optimum way within the current task context.

Paul, 
*  what's your opinion of making a distinction between tasks that are intended to be followed-up and tasks that are intended to
  be used as starting base for the creation of followed-up tasks ?
*  we should maximum re-use current available or already planned possibilities. How will the task server work ? Which role can it 
  fullfill for this feature request ?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-02-12T12:35:03Z says:_ 

 Peter, thanks for responding, and describing in detail how you do this.

Even though we call this by different names, and expect it to work differently, I think we both want the same thing.  I will still argue every detail with you, but that's not really important.  I would like this feature.  Here is my usage scenario, and I think I have a simpler solution than you, and it achieves the same thing.

Every year I plant tomatoes.  There are several tasks involved:

* Prepare soil mix
* Plant seeds
* Thin seedlings
* Move the plants into the greenhouse
* Re-pot the plants
* Move the plants outside
* Stake the plants
* Empty pots, compost the plants
* Store pots for the winter

These 9 tasks have dependencies.  They have due dates.  I repeat this cycle of tasks every year.  Now, let's say I have a single file, tomatoes.txt.  It would look like this:

<pre>
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-04-01 Prepare soil mix
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-04-07 depends:last Plant seeds
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-05-01 depends:last Thin seedlings
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-05-15 depends:last Move the plants into the greenhouse
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-06-15 depends:last Re-pot the plants
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-07-01 depends:last Move the plants outside
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-07-01 depends:last Stake the plants
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-10-01 depends:last Empty pots, compost the plants
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-10-15 depends:last Store pots for the winter
</pre>

Now I just made up this data, which might explain why my tomato crops are disappointing.  Then every year I would run:

<pre>
$ task import tomatoes.txt
</pre>

This would require that the import command recognize 'YYYY' in the due date as a placeholder for the current year.  Also "depends:last" would be recognized as "the most recently entered task".

Once nice aspect of this is that I have a template (repository in your parlance) that represents some recurring project, and being a single file, could be placed on the website as a downloadable thing, like a color theme, or a holiday file.  These are a kind of workflow template.

Now, to answer your questions:
1.  I'm not enthusiastic about creating a new kind of task that is intended as a base, but then perhaps that is exactly what I just described above.  I'd prefer a clever use of the current system than creating a new special case.  This year I am trying to shrink the taskwarrior core code as much as I can.
1.  The task server is all about list sharing and synchronization.  It could help here.  How does it work?  It's basically online storage with a multi-client, multi-user support, that will ultimately allow, for example, you to create a task and delegate it to me, provided I have granted you that privilege.  It will support groups of users with common lists.  It will also have server-side plugins that have infinite capabilities.

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-13T12:25:07Z says:_ 

 Paul, we indeed have the same goal and we both want this feature ! I'm looking forward in discussing the pro's and cons of each other ideas.

I find that the file proposal has following disadvantages:
*  files must be created manually. Not every user will be happy to do that
*  a file layout need to be agreed
*  a file 'language' (YYYY, depends:last,...) is required

Suppose instead of repository we call it a task database (task database is already in use terminology) and we foresee the ability within task to work with multiple databases. Multiple database support is a much more global option that also can be used for our goal. Multiple database usage options:
*  a database for home tasks and a database for work tasks
*  a database for active tasks and template tasks
*  .....

In the FAQ following question is also related: How can I separate my work tasks from my home tasks? Specifically, can I keep them completely separate?

Each database can be managed using the standard task commands.

For example for the tomatoes:

task create db:tomatoes (creates a new data.location=/home/peter/.task/tomatoes/ in .taskrc)
task switch db:tomatoes (from now on each task command is executed on the tomatoes task database)
The different tomatoes tasks can be created using the standard task commands (incl. recurrence, dependencies,....)
Databases can be synchronized using the standard task sync commands
Tasks can be imported/exported using the standard task import/export commands 

Required developments:
*  new task command for the creation of a database
*  new task command for the listing of the available databases
*  enhance some existing commands (task duplicate, task update,....) so they support multiple databases
 
Suppose you've created:
*  the template tomatoes database
*  the recurrent and dependent tomatoes tasks in the template tomatoes database
and a new year start one could then use task <filter> update ... to copy the required tomatoes tasks for this year from the template tomatoes database to your active task database in which you can follow-up the execution of those tasks

What do you find of this solution direction ?

**Comment by taskwarrior** [comment]
_steve rader on 2011-02-14T08:21:02Z says:_ 

 > What do you find of this solution direction ?

I think your suggestions are slightly confused: is the desire to have N
task data stores or is the desire to automate the creation of task lists by
cloning sets of tasks?  You seem to be talking about both.

IMHO, the TASKDATA/TASKRC shell variables [1] address (or at least start to
address) the former and Paul's previous post elegantly addresses the latter.

I think Paul's ideas ought to be broken down into some import enhancements
and perhaps your ideas ought to be broken down into some enhancements for
task-native commands for handling N data stores.

I don't think "task create" is necessary because task automatically creates
data stores. Commands for switching and listing data stores may be a good 
thing--although I don't know how the jive with the task 2.0+ vision(s) and I 
wonder: are they really that much more useful than say "ls -d ~/.task*" and 
say "export TASKDATA=~/.task-grow-tomatoes"?

[1] http://taskwarrior.org/issues/show/632

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-14T14:03:53Z says:_ 

 I'm searching an easy implementation method that allows easy re-use of task lists. The most natural thing is to allow the user creating re-usable task lists in the same way as normal task lists. However most users are going to want to have a clear separation between the tasks they manage and the tasks they use to copy from. Those re-usable task lists must be saved somewhere. The most natural location is another task database. So I come to the task multi database requirement. If task should have the multi database possibility users could also use it for other task organisation needs.  

shell variables, ls, export,... aren't for me really user-friendly solutions. I'm searching for a more in task integrated solution.

**Comment by taskwarrior** [comment]
_Federico Hernandez on 2011-02-15T02:48:35Z says:_ 

 I think all the described above shows that this functionality is really not (and should not be) part of task's core. Instead all the mentioned steps above (like copying template tasks via vim or import) could be nicely put together into a "smart" script that would achieve the wanted result without having to put anything into task.

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-15T10:40:34Z says:_ 

 It's not because it can be implemented via a smart script that it isn't task core functionality. 
The key question is what are the objective decision criteria to decide if something belongs to the task core functionality or not.
For me it's core functionality for a task management program.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-02-15T11:54:52Z says:_ 

 There are two definitions of "core" here.  We're having our standard terminology argument.

Peter is defining "core" as "essential to the business of task management", and I agree with his statement in that context.  Peter is talking about the domain core.

Federico and I are using "core" to mean "written in C++ and compiled into the taskwarrior binary".  In that respect, if there is something that can be achieved by an external script (external to the core, that is), then it is a good idea, because it maintains or shrinks the core, while adding functionality.  We are talking about the software core.

A good example is the set of "export" commands.  They are currently in core, and the number of supported formats keeps growing.  This is bad, because it means the core will never be complete.  Going forward, these export commands will be scripted, and use the _query helper command, which emits only JSON.  Thus the export scripts are now simply filters that transform JSON into the desired format.  It now becomes much simpler for someone to obtain their export format of choice, because it can be achieved by taking one of our reference implementations in Perl, Ruby and Python, and changing them accordingly.  These scripts are now about 20 lines of code, making new export formats both downloadable, and something the community can more readily contribute.

Notice also how with the advent of custom reports, the tabular reports have been migrated out of core.

It is not possible to drive any product to completion, if one aspect of that product is to support "fashionable" features, such as export formats, report settings and so on.  This way, we have at least a chance at achieving completion of the core.

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-15T14:38:10Z says:_ 

 Paul, based on above info and taking into account the desired 'core protection' what's now your preferred solution direction for going forward ?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-02-15T15:05:47Z says:_ 

 Hi Peter,

I like this idea, and consider it an important capability we should add to taskwarrior.

Your approach is more comprehensive, and allows for manipulation of these databases without editing.  My approach was more lightweight, but requires editing the databases.  Unless a third option comes up, we'll be mixing and matching from these options, based on this ongoing discussion.  I don't think we're done yet.

I'm neutral on exactly how this ends up, although I would lean toward the option that involves less work for me.

I'm shrinking the core C++ code.  That means whatever we come up with will be implemented as a plugin, or an external script or program.  This is largely irrelevant, because it will be indistinguishable to the user.

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2011-02-15T15:49:41Z says:_ 

 Paul, thanks for the feedback.  

Suppose someone writes a python addon which are the offered integration possibilities ? Is there a API(module) that can be used from python or is directly triggering of task commands the only option ?

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-02-15T15:51:48Z says:_ 

 Right now, the only option would be calling existing commands.  Soon (2.x) this could be achieved via the Lua API.

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2011-05-06T07:38:10Z says:_ 

 This is a feature I've long wanted in a task management tool.  :-)  I've always thought of it as "task templates"; "repository" seems like it would be odd terminology.

I personally wouldn't mind editing a text file to set up templates.  Indeed, that'd make it fairly straightforward to review, update, and customize templates, and to share them with my teammates or co-workers easily.

The templates need to support some form of keyword substitution.  For instance, if I used Paul's gardening template for tomatoes, carrots, and beans, I'd end up with 3 copies of each task that had identical titles, which could get confusing in reports where pro is omitted.  So we need to be able to define the task template like this:

project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-04-01 Prepare soil mix for [PLANT] planting
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-04-07 depends:last Plant [PLANT] seeds
project:tomatoes rc.dateformat:YYYY-MM-DD due:YYYY-05-01 depends:last Thin [PLANT] seedlings
...

Then I would run it like:

$ task import tomatoes.txt
Error:  Please specify PLANT=<text>

$ task import tomatoes.txt PLANT=tomato
$ task import tomatoes.txt PLANT=carrot
$ task import tomatoes.txt PLANT=bean

**Comment by taskwarrior** [comment]
_Bryce Harrington on 2011-05-06T07:54:08Z says:_ 

 A benefit to implementing this as an extension outside core is that this could make it more easily extended.

For instance, imagine a sysadmin that is going to be doing upgrades to a collection of computers, one per week.  He wants to customize the task lists for the types of software running on the hardware, and also wants to include additional tasks for the older hardware.  So he wants to customize the extension to build the task list by querying his machine database and generating extra tasks based on various criteria.  He also mods the extension to allow him to set waits on each set of tasks to space out one machine per week and to space out the tasks through the week.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-05-07T02:49:19Z says:_ 

 Yes, it's a good idea.  We just need to solve a few outstanding minor details, no technical hurdles.  Once this converges, we can add it.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2017-01-17T02:55:34Z says:_ 

 5 years, no activity.
